DAOS-4233 debug: allow to limit log file footprint

bfaccini opened the pull request:
The idea here is to avoid DAOS log files to become so huge
that they can fill the filesystem where they stand.
One of the usecases is that this is particularly likely to
occur during CI tests running on nodes with poor disk
resources or more easily on VMs.

This patch is a port+merge of PR-2679 from Liang, it
includes the following changes:

  _ Add a limit for debug log file size (default is 1GB)

  _ When to log size exceeds this limit, it will be renamed
to ${D_LOG_FILE}.old. A new file will be created and become
the active log file

  _ D_DEBUG() does not write the logfile most of the time.
It only copies log message into log buffer by default, it writes
to the logfile only if the log buffer is full or exceeds the
predefined write interval (1 second).

  _ D_ASSERT flushes the log buffer and sync the log file.

It also makes CI scripts aware so that .old log files will be
saved as artifacts.

Change-Id: I9e892e24e5c945a00da4c73e348ef6403ec4f7da
Signed-off-by: Bruno Faccini <bruno.faccini@intel.com>
Signed-off-by: Liang Zhen <liang.zhen@intel.com>

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/1/execution/node/60/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/2/execution/node/60/log

brianjmurrell commented:
> On of th usecases is that this is particularly likely to

Two spelling errors in the above line of the commit message.

> occur during CI tests running on nodes with poor disk
> resources

@bfaccini Just to be clear, this issue has nothing at all to do with CI.  CI has 30GB of space for logs.  DAOS should just not be creating logs so large that they fill 30GB of space.  Thus, CI is merely a demonstration of how big and low-value (in reference to the space they can consume -- many 10s of GBs) the logs are.  Big logs are a problem everywhere including both CI and production environments.

> or more easily on VMs.

VMs is a complete strawman here.  Whether a cluster is VM or hardware has absolutely no bearing on log size and management.

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/3/execution/node/61/log

bfaccini commented:
> There is a log cleanup ...

Well, right !! It would be a shame to end up filling the CI nodes with these .old files !! ;-) Thanks to point this.

brianjmurrell commented on doc/environ.md:
```diff
@@ -108,6 +108,16 @@ DAOS debug logs (both server and client) are written to /tmp/daos.log by
+### `D_LOG_SIZE`
+
+DAOS debug logs (both server and client) have a 1GB file size limit by
+default. When this limit is reached, current log file is closed and
```
```suggestion
default. When this limit is reached, the current log file is closed and
```

brianjmurrell commented on doc/environ.md:
```diff
@@ -108,6 +108,16 @@ DAOS debug logs (both server and client) are written to /tmp/daos.log by
+DAOS debug logs (both server and client) have a 1GB file size limit by
+default. When this limit is reached, current log file is closed and
+renamed with a .old suffix, and a new one is opened. This mechanism
+will repeat each time limit will be reached, meaning that available
```
```suggestion
will repeat each time the limit is reached, meaning that available
```

brianjmurrell commented on doc/environ.md:
```diff
@@ -108,6 +108,16 @@ DAOS debug logs (both server and client) are written to /tmp/daos.log by
+will repeat each time limit will be reached, meaning that available
+saved log records could be found in both ${D_LOG_FILE}[.old] files.
+This can be modified by setting this environment variable
+("D_LOG_SIZE=536870912").
```
Is it worth mentioning that only the last generation of `.old` is kept such that anything older than `2*D_LOG_SIZE` is discarded and lost?  Just so people are not confused about the earliest log records being missing.

bfaccini replied:
Well, any details are welcome on the doc side, what is really limiting me in this domain is that I have to do it in English ;-)

brianjmurrell replied:
Indeed.  I can most certainly appreciate that.  Your English is **far** better than my French.

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/4/execution/node/61/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/5/execution/node/61/log

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/5/execution/node/503/log

jolivier23 commented on src/include/gurt/debug.h:
```diff
@@ -279,8 +280,10 @@ int d_register_alt_assert(void (*alt_assert)(const int, const char*,
-			D_FATAL("Assertion '%s' failed\n", #e);	\
+		if (!(e)) {						\
+			D_FATAL("Assertion '%s' failed\n", #e);		\
+			d_log_sync();					\
```
what about sigsegv or other such signals?   I also would want the log dumped in such cases

jolivier23 commented on src/gurt/dlog.c:
```diff
@@ -591,11 +763,19 @@ d_log_open(char *tag, int maxfac_hint, int default_mask, int stderr_mask,
 		truncate = 1;
 
+	env = getenv(D_LOG_SIZE_ENV);
+	if (env != NULL) {
```
we could probably add a function like d_getenv_size that takes into account K, M, and G like ts_val_factor in daos_perf.c and make this a little more friendly to set (e.g. 1G rather than 1000000000 or whatever).

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/6/execution/node/60/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

jolivier23 reviewed (dismissed):
I don't see any major issues.    Wondering how this will play with all of the log parsing code though. 
 @ashleypittman

jolivier23 commented:
Maybe I answered my own question.   There are new warnings in the NLT

https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/7/VM_test/new/

bfaccini commented:
> Maybe I answered my own question. There are new warnings in the NLT
> 
> https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/7/VM_test/new/

Well I have been there, but it looks pretty obscure to me ... And BTW, the indicate warnings locations don't seem related to this change.

ashleypittman commented:
I'll be clear, I think this is the wrong solution and doesn't really solve anything.  It looks like the default is 1Gb per file, then duplicated so 2Gb is the most logging/debugging we can have with default settings.  Is there an ability to turn it off for specific test runs?

It shouldn't affect the nlt stuff, because as far as I'm aware the largest log that we get from that is around 64Mb, the extra errors are some apparent memory leaks in a part of the code that as Bruno says is unchanged, but they could also be from a truncated log file which this patch could be responsible for.

There is potentially a big performance improvement here because it's effectively logging to a ring buffer and syncing to disk when the buffer is full so there are fewer I/O requests, although this in itself will cause a problem where multiple processes are writing to the same log file which does not appear to have been considered.

What I'd like to see as a minimum is the test code being updated to check for the presence of any "log.old" files, and automatically failing any test where they were created, so that we both don't lose any logs in CI but also so that we know which tests are doing this and can tune the settings for them automatically.

bfaccini commented:
> I'll be clear, I think this is the wrong solution and doesn't really solve anything. It looks like the default is 1Gb per file, then duplicated so 2Gb is the most logging/debugging we can have with default settings. Is there an ability to turn it off for specific test runs?

Well you can specify a new limit of exabytes using D_LOG_SIZE env var.

> There is potentially a big performance improvement here because it's effectively logging to a ring buffer and syncing to disk when the buffer is full so there are fewer I/O requests, although this in itself will cause a problem where multiple processes are writing to the same log file which does not appear to have been considered.

It is true if you think to maintain the exact time-ordered sequence, we may need to have a post-processing step to re-order msgs between 16K buffered chunks ? File-switching operations are also subject to collisions in this case, so we may use a semaphore/IPC or a Futex to synchronise these ops ? What do you think? In all cases, it could be easily avoided by using D_LOG_FILE_APPEND_PID.

> What I'd like to see as a minimum is the test code being updated to check for the presence of any "log.old" files, and automatically failing any test where they were created, so that we both don't lose any logs in CI but also so that we know which tests are doing this and can tune the settings for them automatically.

Hum, I think current CI tests failures are already pointing to some changes required, like for the ZeroConfigTest test where it seems that log file parsing used for this test needs to handle the .old one.

brianjmurrell commented:
> I'll be clear, I think this is the wrong solution and doesn't really solve anything. It looks like the default is 1Gb per file, then duplicated so 2Gb is the most logging/debugging we can have with default settings. Is there an ability to turn it off for specific test runs?

I think what you are looking for is in the PR: https://github.com/daos-stack/daos/blame/5267ce8da173a36fe6fd468649c1f37b2b936dbb/doc/environ.md#L111-L121

> What I'd like to see as a minimum is the test code being updated to check for the presence of any "log.old" files, and automatically failing any test where they were created, so that we both don't lose any logs in CI but also so that we know which tests are doing this and can tune the settings for them automatically.

This sounds like the work that a PR that @amjustin13 is working on currently would do.

ashleypittman commented:
> > I'll be clear, I think this is the wrong solution and doesn't really solve anything. It looks like the default is 1Gb per file, then duplicated so 2Gb is the most logging/debugging we can have with default settings. Is there an ability to turn it off for specific test runs?
> 
> I think what you are looking for is in the PR: https://github.com/daos-stack/daos/blame/5267ce8da173a36fe6fd468649c1f37b2b936dbb/doc/environ.md#L111-L121

That would allow you to tune the size but not disable it?  How about a value of 0 for off?

> > What I'd like to see as a minimum is the test code being updated to check for the presence of any "log.old" files, and automatically failing any test where they were created, so that we both don't lose any logs in CI but also so that we know which tests are doing this and can tune the settings for them automatically.
> 
> This sounds like the work that a PR that @amjustin13 is working on currently would do.

Yes, I'm aware of work towards this.  It seems this PR and that one would need to be integrated properly.

I'm concerned with this PR that it effectlly allows us to do unlimited logging as it just gets thrown away as part of the test but the net result being that daos is really verbose with low quality messages, and whenever there's a problem in CI (or elsewhere) it'll be really difficult to find anything because it'll either be hidden or deleted.

ashleypittman commented:
> > I'll be clear, I think this is the wrong solution and doesn't really solve anything. It looks like the default is 1Gb per file, then duplicated so 2Gb is the most logging/debugging we can have with default settings. Is there an ability to turn it off for specific test runs?
> 
> Well you can specify a new limit of exabytes using D_LOG_SIZE env var.
> 
> > There is potentially a big performance improvement here because it's effectively logging to a ring buffer and syncing to disk when the buffer is full so there are fewer I/O requests, although this in itself will cause a problem where multiple processes are writing to the same log file which does not appear to have been considered.
> 
> It is true if you think to maintain the exact time-ordered sequence, we may need to have a post-processing step to re-order msgs between 16K buffered chunks ? File-switching operations are also subject to collisions in this case, so we may use a semaphore/IPC or a Futex to synchronise these ops ? What do you think? In all cases, it could be easily avoided by using D_LOG_FILE_APPEND_PID.

The improved performance is to me the best part of this PR, and I'm OK with removing the time-ordered sequence to achieve that, but in the past it's also been useful and I know traditionally most of our tests just logged to a single file.  I assume we just open the file with O_APPEND and the kernel avoids splicing any writes.  I'd have to check the PR but if this flushes entire lines at a time rather than a fixed number of bytes that would avoid the worst cost as at least we could sort the logs and know we weren't getting any splicing.

> > What I'd like to see as a minimum is the test code being updated to check for the presence of any "log.old" files, and automatically failing any test where they were created, so that we both don't lose any logs in CI but also so that we know which tests are doing this and can tune the settings for them automatically.
> 
> Hum, I think current CI tests failures are already pointing to some changes required, like for the ZeroConfigTest test where it seems that log file parsing used for this test needs to handle the .old one.

Do you have a link?  Yes, the log file parsing should also be updated to handle the .old component, as well as check the two files related to the same pid.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/d40c16eda08fb4312cf6a65677147033b9f8093e/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/8/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/8/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/8/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/8/display/redirect

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/d40c16eda08fb4312cf6a65677147033b9f8093e/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

ashleypittman commented:
> LGTM. No errors found by checkpatch.
> 
> FYI: Errors found in lines not modified in the patch:
> 
> [src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/d40c16eda08fb4312cf6a65677147033b9f8093e/src/tests/ftest/network/zero_config.py#L32):
> (pylint-unused-import) Unused include_local_host imported from agent_utils

I know it's unrelated to this PR but this seems like a simple one to fix if you're re-spinning.

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/9/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/9/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/9/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/9/display/redirect

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/d40c16eda08fb4312cf6a65677147033b9f8093e/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/10/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/10/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/10/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/10/display/redirect

johannlombardi commented:
I would like to complete the review cycle and land this PR. I agree that this is not the perfect solution and we will be looking next into more advanced one. Meanwhile, we need to do something and this PR seems reasonable to me as a first step.

ashleypittman commented:
> Meanwhile, we need to do something and this PR seems reasonable to me as a first step.

I've been out but is are there specific issues that need resolving currently?  I know we were having issues in CI where logs were filling all available space and therefore not getting copied/saved appropriately however we're improved this by reducing the log level in some tests, changing the out-of-space behaviour of the test framework and simply turning down severity of some messages or removed them entirely.

brianjmurrell commented:
@ashleypittman I suppose if we were to decide that 2G is the limit of the size of log files that we want to produce (which it seems we have with this PR), this PR **enforces** that decision, and that as long as useful logs are being created under that threshold then this PR doesn't cause any harm.

If we are finding that 2G is not enough to get useful logs, we can continue to investigate and eliminate noise in the logs until we get it under the 2G and again this PR is causing no harm.

brianjmurrell commented on doc/environ.md:
```diff
@@ -108,6 +108,18 @@ DAOS debug logs (both server and client) are written to /tmp/daos.log by
+generation of ${D_LOG_FILE}.old files, to a maximum of the most recent
+2*D_LOG_SIZE records.
+This can be modified by setting this environment variable
+("D_LOG_SIZE=536870912").
```
Should add a comment mentioning that human-readable sizes (i.e. 1G) can be used here?  Something like:

```suggestion
("D_LOG_SIZE=536870912").  Sizes can also be specified in human-readable form using `k`, `m`, `g`, `K`, `M`, and `G`.  The lower-case specifiers are base-10 multipliers and the upper case specifiers are base-2 multipliers.
```

bfaccini replied:
Ok, agree! BTW, I need to move this to doc/admin/env_variables.md as doc/environ.md has recently been removed

brianjmurrell commented on src/tests/ftest/network/zero_config.py:
```diff
@@ -143,6 +143,9 @@ def verify_client_run(self, exp_iface, env):
         racer_env["FI_LOG_LEVEL"] = "info"
+        # bump/double log size limit to avoid log switch and be able to
+        # find initialization msg in single/first log file
+        racer_env["D_LOG_SIZE"] = "2G"
```
Is this the correct fix here or should the test be aware of the `.old` log possibility and check that log (first, since it's looking for the beginning of the log, isn't it?) if it exists?

bfaccini replied:
This is the simplest fix! And anyway, raising the size may become necessary if the log verbosity raises and will allow to exceed 2x1GB ...

brianjmurrell replied:
> This is the simplest fix! 

Indeed, it is the simplest.  But simplest is not always the correct fix.

> And anyway, raising the size may become necessary if the log verbosity raises and will allow to exceed 2x1GB ...

At which point I'd say that the test failing is just a symptom of the larger problem of having excessively large logs with low quality.  At that point if large log files is causing that test to fail, the fix is to address the log size, not just keep allowing them to be bigger and bigger, just to pass tests.

bfaccini replied:
> At that point if large log files is causing that test to fail, the fix is to address the log size

You mean the reason why the log is growing too much ? Then this is not the object of this ticket/PR right ?

brianjmurrell replied:
Yes.  This PR isn't to address reasons why the log grows, but just to limit the size of a log and the ripple effects caused by excessive logs.  But what I am saying is that if the log grows beyond the default limit of 2G (1G + 1G .old) and that affects a test (which reads the log, say) or debugging efforts, then something needs to be done to address the size (i.e. quality) of the log.

But getting back to reason for this conversation, the problem with the "fix" https://github.com/daos-stack/daos/blob/d40c16eda08fb4312cf6a65677147033b9f8093e/src/tests/ftest/network/zero_config.py#L148
is that test is reading the log and wants to read what's at the beginning of it.  But due to this PR, the beginning of the log could be in the `.old` log file.  So the fix for that is to make the test know that it needs to read from the `.old` log file, if it exists, rather than just allowing the test to create bigger logs so that a `.old` is not generated.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/184077e80021c8950bfa29e750f84edd367502c6/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

brianjmurrell reviewed (changes requested):
Test that reads the log file is still not changed to read the `.old` log file if it exists which is where it will find the information it needs.

jolivier23 commented on doc/admin/env_variables.md:
```diff
@@ -68,6 +68,7 @@ Environment variables in this section only apply to the client side.
 |Variable    |Description|
 |------------|-----------|
 |D\_LOG\_FILE|DAOS debug logs (both server and client) are written to `/tmp/daos.log` by default. The debug location can be modified by setting this environment variable ("D\_LOG\_FILE=/tmp/daos_debug.log").|
+|D\_LOG\_SIZE|DAOS debug logs (both server and client) have a 1GB file size limit by default. When this limit is reached, the current log file is closed and renamed with a .old suffix, and a new one is opened. This mechanism will repeat each time the limit is reached, meaning that available saved log records could be found in both ${D_LOG_FILE} and last generation of ${D_LOG_FILE}.old files, to a ...
```
we should get out of the habit of putting things on long lines.  The MD parser will remove the new lines

johannlombardi replied:
Let's fix that in a subsequent doc-only PR.

jolivier23 replied:
@johannlombardi agreed.   Just making a general comment as it makes reviewing doc changes impossible

jolivier23 commented on src/gurt/dlog.c:
```diff
@@ -376,6 +549,7 @@ void d_vlog(int flags, const char *fmt, va_list ap)
 	 * errno to its original value
 	 */
 	int save_errno = errno;
+	int rc;
```
Not sure what happened to my comment so I'm making it again.  Below, we do clog_lock and then proceed to do gettimeofday.  I think we should do the gettimeofday outside of the locked region to minimize the exclusive period.   Also, we might consider using something other than gettimeofday such as the gurt time routines that use clock_gettime instead as I believe gettimeofday is a system call.   Particularly if we are mostly avoiding system calls except on buffer flushes.

bfaccini replied:
Well, to me both clock_gettime() and gettimeofday() should be similar and not of a high cost since they are likely to be handled in vdso code on most architectures.

bfaccini replied:
@jolivier23 do you agree with my previous comment ? Also last CI run is reported as having failures, but all tests seem to have been successful....

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/184077e80021c8950bfa29e750f84edd367502c6/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

johannlombardi commented:
Unit test are failing. Is that expected?

bfaccini commented:
> Unit test are failing. Is that expected?

NLT unrelated errors ?

ashleypittman commented:
> Unit test are failing. Is that expected?

No, it's not expected.  Details are here https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/13/VM_test/new/

It looks like memory leaks however looking at the log file at https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/13/artifact/unit_vm_test/dnt_cmd_32_zxun4hlr.log/*view*/ I wonder if perhaps the log file just isn't flushed on this exit path?

Either way it's a genuine regression.

daosbuild1 commented on src/gurt/dlog.c:
```diff
@@ -701,6 +932,14 @@ d_log_open(char *tag, int maxfac_hint, int default_mask, int stderr_mask,
+	 */
+	rc = atexit(d_log_sync);
+	if (rc != 0)
+		fprintf(stderr, "unable to register flush of log, potential risk to miss last lines if fini method not invoked upon exit\n");
```
(style)  line over 80 characters

daosbuild1 replied:
(style)  line over 80 characters

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/14/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/1950f35071ea701eda1894677edb55555ae46069/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/14/execution/node/60/log

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/14/execution/node/505/log

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/15/execution/node/196/log

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/16/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/db4ab9f0dc8f0200e9cde9e3b56abd0c720ceea6/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/16/execution/node/60/log

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/16/execution/node/505/log

daosbuild1 commented on src/gurt/dlog.c:
```diff
@@ -701,6 +936,14 @@ d_log_open(char *tag, int maxfac_hint, int default_mask, int stderr_mask,
+	 */
+	rc = atexit(d_log_sync);
+	if (rc != 0)
+		fprintf(stderr, "unable to register flush of log, potential risk to miss last lines if fini method not invoked upon exit\n");
```
(style)  line over 80 characters

daosbuild1 replied:
(style)  line over 80 characters

daosbuild1 replied:
(style)  line over 80 characters

daosbuild1 replied:
(style)  line over 80 characters

daosbuild1 replied:
(style)  line over 80 characters

daosbuild1 replied:
(style)  line over 80 characters

daosbuild1 replied:
(style)  line over 80 characters

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/17/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/992bcd0d47728098b9d7162d34eb332ce885d64d/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/17/execution/node/60/log

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/17/execution/node/505/log

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/17/display/redirect

ashleypittman commented:
This is still failing Unit Test, most likely with valgrind errors although these haven't been reported yet.  There were some errors on master until last night so it might be a simple rebase but I can't tell until the results are there.

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/18/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/4eea7c7bc1fedfe763dece6e8237b50756f61837/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/18/execution/node/60/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/18/execution/node/983/log

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/19/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/482de5205f0a17f433ee5bed0149ac4c3ad75003/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/19/execution/node/60/log

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/19/display/redirect

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/19/execution/node/861/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/19/execution/node/980/log

bfaccini commented:
Last CI session (#19) failures report :: 
               _ Full “Functional on CentOS 7“ stage failure due to “One or more nodes failed post-provision configuration!“ reason when it has been successful in previous CI session for their PR before to be re-run…
               _ 24 tests in “Functional Leap 15” stage, all due to “Test died without reporting the status“ reason —> current known issue ??
               _ 1st test of DaosCoreTestRebuild serie (test r_0_10) timeout (wait for rebuild pool) during “Functional_Hardware_Medium“ stage  —> already known and addressed in DAOS-5708 JIRA ticket
               _ Single “CSumErrorLog.test_csum_error_logging;“ failure (missing “admin.crt” certificat to run dmg command) during “Functional_Hardware_Medium“ stage  —> already known and addressed in DAOS-5745 JIRA ticket

brianjmurrell reviewed (changes requested):
https://github.com/daos-stack/daos/blob/482de5205f0a17f433ee5bed0149ac4c3ad75003/src/tests/ftest/network/zero_config.py#L90 needs to be updated to handle the fact that the log could be a `.old` log file with something like (untested):

```
        cmd = "if [ -e {0}.old ]; then head -50 {0}.old; else head -50 {0}; fi".format(log_file)
```

rather than allowing a log to grow beyond the set limit https://github.com/daos-stack/daos/blob/482de5205f0a17f433ee5bed0149ac4c3ad75003/src/tests/ftest/network/zero_config.py#L148

bfaccini commented:
> https://github.com/daos-stack/daos/blob/482de5205f0a17f433ee5bed0149ac4c3ad75003/src/tests/ftest/network/zero_config.py#L90
> 
> needs to be updated to handle the fact that the log could be a `.old` log file with something like (untested):
> ```
>         cmd = "if [ -e {0}.old ]; then head -50 {0}.old; else head -50 {0}; fi".format(log_file)
> ```
> 
> rather than allowing a log to grow beyond the set limit
> 
> https://github.com/daos-stack/daos/blob/482de5205f0a17f433ee5bed0149ac4c3ad75003/src/tests/ftest/network/zero_config.py#L148

Ok ..., but I will keep the size bump to have a head start in anticipation of a possible future increase in the volume of the log ;-)

daosbuild1 commented on src/tests/ftest/network/zero_config.py:
```diff
@@ -87,7 +87,8 @@ def get_log_info(self, hosts, dev, env_state, log_file):
         """
-        cmd = "head -50 {}".format(log_file)
+	# anticipate log switch
+        cmd = "if [ -e {}.old ]; then head -50 {}.old; else head -50 {}; fi".format(log_file)
```
(pylint-line-too-long) Line too long (93/80)
(pylint-too-few-format-args) Not enough arguments for format string

brianjmurrell replied:
I don't think you can re-use a single format argument with indexing it:

```
$ python -c 'print("{} {} {}".format("foo"))'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
IndexError: tuple index out of range
$ python -c 'print("{0} {0} {0}".format("foo"))'
foo foo foo
```

```suggestion
        cmd = "if [ -f {0}.old ]; then head -50 {0}.old; else head -50 {0}; fi".format(log_file)
```

bfaccini replied:
oops, I thought index 0 would be the default, particularly when there is only one parameter !!... Python is sometimes surprisingly very "fussy" !!...

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/21/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/a2ac9296064284387b34d524f5e18449939e57f9/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/21/execution/node/60/log

brianjmurrell commented:
> Ok ..., but I will keep the size bump to have a head start in anticipation of a possible future increase in the volume of the log ;-)

I disagree with this.  If this is ever needed, (2G .log + 2G .log.old) in order to still find the beginning of the logging then the logs are too big, by definition of this very PR.  This PR is asserting that 2G of logs (total, 1G `.log` and 1G `.log.old`) should be enough and that if it is not, then the quality of the log is low and that needs to be addressed to make this test pass again, not simply ever increasing the log size until we are at a point of filling disks again.  IMHO.

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/21/execution/node/866/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/21/execution/node/985/log

daosbuild1 commented on src/gurt/dlog.c:
```diff
@@ -701,6 +936,14 @@ d_log_open(char *tag, int maxfac_hint, int default_mask, int stderr_mask,
+	 */
+	rc = atexit(d_log_sync);
+	if (rc != 0)
+		fprintf(stderr, "unable to register flush of log, potential risk to miss last lines if fini method not invoked upon exit\n");
```
(style)  line over 80 characters

johannlombardi replied:
I think we need to specify here somehow that this is coming from the DAOS library.

bfaccini replied:
oops sorry I have missed this comment. Since log initialisation stuff has been done at this point, we may use D_WARN() ?

daosbuild1 commented on src/tests/ftest/network/zero_config.py:
```diff
@@ -87,7 +87,9 @@ def get_log_info(self, hosts, dev, env_state, log_file):
-        cmd = "head -50 {}".format(log_file)
+	# anticipate log switch
+        cmd = "if [ -f {0}.old ]; then head -50 {0}.old; else head -50 {0}; fi"
+              .format(log_file)
```
(pylint-syntax-error) unexpected indent

johannlombardi replied:
indenting looks wrong here, no?

bfaccini replied:
Oops, and I have forgotten the continuation char

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/22/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/22/execution/node/60/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/22/execution/node/202/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/22/execution/node/256/log

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/23/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/15e686f928552d202e134fccb32ff1a32d5ef9c9/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/23/execution/node/60/log

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/23/execution/node/866/log

daosbuild1 commented on src/tests/ftest/get_remote_files.sh:
```diff
@@ -0,0 +1,324 @@
+    # shellcheck disable=SC2086
+    ftp="${FILES_TO_PROCESS%\"}"
+    ftp="${ftp#\"}"
+    files=$(ls -d ${ftp})
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on src/tests/ftest/launch.py:
```diff
@@ -35,7 +34,10 @@
+import xml.etree.ElementTree as ET
+import glob
 
+from xml.dom import minidom
```
(pylint-ungrouped-imports) Imports from package xml are not grouped

daosbuild1 commented on src/tests/ftest/launch.py:
```diff
@@ -885,32 +904,146 @@ def clean_logs(test_yaml, args):
+    """
+    print("Compressing files in {}".format(socket.gethostname().split(".")[0]))
+    logs_dir = os.path.join(avocado_logs_dir, "latest", "daos_logs", "*.log*")
+    # Debug check if this directoy exisits
```
```suggestion
    # Debug check if this directory exists
```

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/24/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/38d55a065d24cbf37c28e0b856362fceb9c06607/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/24/execution/node/61/log

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/24/execution/node/841/log

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/24/execution/node/867/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/24/execution/node/944/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/24/execution/node/987/log

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3405/25/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/network/zero_config.py:32](https://github.com/daos-stack/daos/blob/15e686f928552d202e134fccb32ff1a32d5ef9c9/src/tests/ftest/network/zero_config.py#L32):
(pylint-unused-import) Unused include_local_host imported from agent_utils

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/25/execution/node/61/log

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/25/execution/node/867/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3405/25/execution/node/986/log

brianjmurrell commented:
> I don't see any major issues. Wondering how this will play with all of the log parsing code though.
> @ashleypittman

Was there an answer to this?  Did the code that does the log parsing account for the `.old` log?

In retrospect, I wonder if the 2 1G log files (`.log` and `.log.old`) should have been concatted back together before any other post-test processing were done on them.  That would have eliminated much of the ancillary change in this PR.  Pity I did not think about this possibility while I was reviewing this PR.  :frowning:

ashleypittman commented:
> > I don't see any major issues. Wondering how this will play with all of the log parsing code though.
> > @ashleypittman
> 
> Was there an answer to this? Did the code that does the log parsing account for the `.old` log?
> 
> In retrospect, I wonder if the 2 1G log files (`.log` and `.log.old`) should have been concatted back together before any other post-test processing were done on them. That would have eliminated much of the ancillary change in this PR. Pity I did not think about this possibility while I was reviewing this PR. 😦

No.  The conclusion was that there are no tests other than one that use this feature so it doesn't make much difference.

In theory you could parse the .log.old file, and then the .log (we could just code this into the log parsing rather than concatenate the files in the tree) but this would still only work if the total log size was between 2G and 4G, anything bigger than that would loose part of the logs entirely.

Obviously any log parsing which misses the start of logging will report lots of frees of unallocated memory, and won't track descriptors properly so will be of limited use (it could still track some memory leaks however)

I do wonder if we shouldn't put some kind of header or preamble in the file to make where it starts, and only do log analysis if this block is present.  If we switched from real timestamps to jiffys past a set point then we could combine it with this maybe.
